File: 백준/백준 1976.cpp
// 백준 1976
// union - find 문제

#include <cstdio>

int n, m;
int parent[205], arr[1005];

int find(int x){
    // 타고 올라가서 반환
    if (parent[x] == x) return x;
    return parent[x] = find(parent[x]);
}

void Union(int x, int y){
    // x를 y아래로
    int px = find(x), py = find(y);
    parent[px] = py;
}

int main(){
    // freopen("input.txt", "r", stdin);
    // input
    scanf("%d", &n); scanf("%d", &m);
    for(int i = 1; i <= n; ++i) parent[i] = i;

    for(int i = 1; i <= n; ++i){
        for(int j = 1; j <= n; ++j){
            int temp; scanf("%d", &temp);
            // 두 도시를 왕복 할 수 있다면 같은 집합으로 예편
            if(temp) Union(i, j);
        }
    }

    for(int i = 0; i < m; ++i) scanf("%d", &arr[i]);
    bool flag = true;

    for(int i = 1; i < m; ++i){ 
        if(find(arr[i - 1]) != find(arr[i])){
            // 같은 집합이 아닌 도시다 -> 한번에 여행 할 수 없다.
            flag = false;
            break;
        }
    }
    
    flag ? puts("YES") : puts("NO");
}

File: 백준/백준 1976.py
# 백준 1976

import sys
# sys.stdin = open("./test/input.txt", "r")

si = sys.stdin.readline

n, m = int(si()), int(si())

parent = [i for i in range(n + 1)]

def find(x : int) -> int:
    if parent[x] == x:
        return x
    parent[x] = find(parent[x])
    return parent[x]

def union(x : int, y : int) -> None:
    # x를 y의 밑으로
    px, py = find(x), find(y)
    parent[px] = py

for i in range(1, n + 1):
    arr = list(map(int, si().split()))
    for j in range(n):
        if arr[j] == 1:
            union(i, j + 1)

plan = list(map(int, si().split()))

flag = True
for i in range(1, m):
    if find(plan[i-1]) != find(plan[i]):
        flag = False
        break

print("YES" if flag else "NO")

